### ISSUE TITLE
Possible vnode data race

### ISSUE THREAD

**Issue by kikht** [osxfuse/osxfuse#824]
Hello! In #299 is was decided that `remove` vnop should acquire both parent and child vnode locks. It was eventually implemented in https://github.com/osxfuse/kext/commit/176e6fbda9d0cd35a05b8651c5311cd180b1d750

It seems to me that the same logic should be applied to `rmdir` and `link` vnops. As far as I can tell right now only child vnode lock is acquired by these ops:
* https://github.com/osxfuse/kext/blob/534773bd3135f04b7f6fb801faa2f8991daf4a6b/osxfuse/fuse_biglock_vnops.c#L265-L267
* https://github.com/osxfuse/kext/blob/534773bd3135f04b7f6fb801faa2f8991daf4a6b/osxfuse/fuse_biglock_vnops.c#L610-L612

Could you please take a look at these functions?

**Comment by bfleischer** [comment]
HFS and NFS lock the vnode itself and the directory vnode in rmdir and link. macFUSE currently only locks the vnode.

Are you running into any issues with the current implementation?

**Comment by kikht** [comment]
I'm experiencing some strange behaviour with directory deletion. Sometimes directories appear existing for some time after successful RMDIR op. Unfortunately, I cannot reproduce this reliably, it only happens under load. Trying to debug this lead me to vnode lock inconsistency.

**Comment by bfleischer** [comment]
Are you seeing these issues using libfuse or a different user space FUSE library?

**Comment by kikht** [comment]
I've seen this problem with both libfuse and my own userspace implementation.

**Comment by kikht** [comment]
It usually happens while deleting large directory trees while having some parallel read operations, e.g. removing directories in IDE while it runs indexing.

**Comment by kikht** [comment]
This race can be reproduced on loopbackfs with the following program. It either can `stat` just removed dir without `ENOENT`, or some `stat` call fails with `ENXIO`. Problem disappears when filesystem server is run in single-threaded mode (`-s`) or without vnode name cache (`-onovncache`). 

```
#include <errno.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>

void remove_safe(const char* path) {
    if (remove(path) != 0) {
        printf("remove failed: %d\n", errno);
        exit(1);
    }

    struct stat buf;
    if (stat(path, &buf) == 0) {
        printf("stat for removed dir is successful\n");
        exit(2);
    }
    if (errno != ENOENT) {
        printf("unexpected error in stat after remove: %d\n", errno);
        exit(3);
    }
}

void* child_thread(void* arg) {
    remove_safe("test_dir/dir");
    return NULL;
}

int main() {
    for (int i = 0; i < 1000000; i++) {
        system("rm -rf test_dir");
        mkdir("test_dir", 0777);
        mkdir("test_dir/dir", 0777);

        pthread_t child;
        pthread_create(&child, NULL, child_thread, NULL);

        while (1) {
            struct stat buf;
            if (stat("test_dir/dir", &buf) != 0) {
                if (errno == ENOENT) {
                    break;
                }

                printf("unexpected error in stat loop: %d\n", errno);
                return 4;
            }
        };

        pthread_join(child, NULL);
    }

    return 0;
}
```

**Comment by bfleischer** [comment]
The fix is in the upcoming 4.2.1 release. Thank you very much for your work.
